### ISSUE TITLE
Using setURLSuffix on URLSegment field is not respected in SiteTreeURLSegmentField.js correctly

### ISSUE THREAD

**Issue by RVXD** [silverstripe/silverstripe-cms#2723]
I'd like to be able  to remove the ?stage=Stage link from the URLSegment field in the CMS. Some clients keep copying this link into their newsletters/socialmedia messages. If people click on it these links they need to login. For most non-technical users this is confusing.

Removing the urlSuffix in Page.php by setting it to false works in source, but javascript adds it again. Resulting in link with undefined at the end.

**Setting URLSuffix to false, if page is published:**
```
public function getCMSFields()
{
      $fields = parent::getCMSFields();
        if( $this->isPublished() ) {
            $fields->FieldByName('Root.Main.URLSegment')->setURLSuffix(false);
        }
}
```
**Workaround setting URLSuffix to '?', if page is published:**
Now all the links will end with a '?'
```
public function getCMSFields()
{
      $fields = parent::getCMSFields();
        if( $this->isPublished() ) {
            $fields->FieldByName('Root.Main.URLSegment')->setURLSuffix('?');
        }
}
```
 **Current SiteTreeURLSegmentField.js** 
Always adds suffix. If it is empty or false causing 'undefined' in the URL.
```
// Transfer current value to holder
this.find('.URL-link').attr('href', encodeURI(url + field.data('suffix'))).text(previewUrl);
```
**Proposed change to SiteTreeURLSegmentField.js:**
Only add suffix if it not false.
```
// Transfer current value to holder
this.find(".URL-link").attr("href",encodeURI(t+(e.data("suffix")?e.data("suffix"):''))).text(n)}
```

**Comment by michalkleiner** [comment]
Do you want to open a pull request, @RVXD?

**Comment by emteknetnz** [comment]
Sorry, we can't really do this.  The ?stage=Stage query string is used to ensure that [HTTP caching is disabled](https://github.com/silverstripe/silverstripe-cms/pull/2471).  Also at once point ?stage=Stage wasn't always used and instead sessions were used to say which stage you were on.  However this was changed to always using the ?stage=Stage to resolve a class of bug of which I can't exactly remember

**Comment by RVXD** [comment]
I still think the change to SiteTreeURLSegmentField.js could and should be made without having a negative impact on the functionality. A lot of users think this is the URL they can communicate to the world. The ?stage=Stage part does not have to be removed at all from  the preview pane or anywhere else. Just from the segment field, so that if someone copies it, it is not in there.
The functionality of $fields->FieldByName('Root.Main.URLSegment')->setURLSuffix(false) is broken in the javascript part of the CMS. If set to false it should not give back undefined in the url.
Hope you can give it another thought.

**Comment by michalkleiner** [comment]
Yeah, I think if @emteknetnz takes a second look and reads through the issue again, it is really only about fixing the JS to correctly respect the already existing settings, nothing else. It's not intended to remove the query param.
I'll reopen the issue and adjust the title.

**Comment by michalkleiner** [comment]
Alternatively, if we want to ensure `false` can't be passed to `setURLSuffix` we need to throw an application exception.

**Comment by GuySartorelli** [comment]
This change won't affect the default behaviour, but makes extending or changing the behaviour to suit the needs of a given project easier. Makes sense to me.

**Comment by purplespider** [comment]
I came here looking to do as @RVXD wants to, but this bug stops it from being possible to remove `?stage=Stage` for those who wish to.

Another workaround for the problem @RVXD is trying to stop could be to override `SiteTreeURLSegmentField.ss` to add a "Copy" button, .e.g. adding:
````html
<button onclick="navigator.clipboard.writeText('{$URLPrefix}{$Value}')" role="button" type="button" class="btn btn-outline-secondary btn-sm font-icon-clipboard-pencil">Copy</button>
````
Which then looks like this:
<img width="519" alt="Screenshot 2024-02-23 at 16 51 44@2x" src="https://github.com/silverstripe/silverstripe-cms/assets/329880/7b0d9882-b85f-4036-8572-515a7e696796">

**Comment by sunnysideup** [comment]
Bump. I also noticed this undefined at the end.

**Comment by patricknelson** [comment]
The root cause pretty simple; improper use of a potentially `undefined` value in JS. @RVXD's proposed fix in `SiteTreeURLSegmentFIeld.js` very safely and directly addresses it, ~~i.e. `e.data("suffix") || ''`~~. Just substitute an empty string if `undefined`. ~~Also, `|| ''` is safe here since [jQuery's `.data()`](https://api.jquery.com/data/) is aligned with `HTMLElement.dataset` and [values are always a string](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset#accessing_values)... if set!~~ 😅 Submitted #3178 to hopefully settle this for good.

Plugging in `->setSuffix('?')` is a workable fix (or `&` in my case for a DataObject URL that already has a query string), but it's _hacky AF_.

p.s. The related reason for why this is happening, and why the more obvious PHP fixes like `->setSuffix('')` don't work, is because `getAttributesHTML` [filters _empty looking_ attributes out](https://github.com/silverstripe/silverstripe-framework/blob/b85cb6ec1e75b7657c2686d20aeb2246868c4229/src/View/AttributesHTML.php#L108-L110), which is a different and potentially very complicated issue. However: This is a perfect example of where you would _want_ an empty string (to compensate for a downstream JS bug), since an empty string is still a valid usable value. Not to mention that the interpretation of attributes varies wildly from other types of data structures. For example, just consider all the ways HTML elements can handle boolean-style attributes!

---

_*Edit: Correction, jQuery doesn't use `dataset` behind the scenes... so a little extra work is needed for a safe fix, but no worries, updating PR._

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/9f794c62-39eb-45a5-9583-c1d69088ed89" />

**Comment by patricknelson** [comment]
Actually, just discovered one really fun workaround: `->setURLSuffix('[]')`

This looks non-empty to PHP and happily passes it along to `data-suffix="[]"`. jQuery then interprets it as an array `[]` instead of a literal string `"[]"`.  When concatenated with the `url` string, that array will then coerce to an empty string. 😂 It's a bug's world, we just live in it.

**Comment by ChloeSartorelli** [comment]
PR merged. This will be automatically tagged by GitHub Actions
